zkwrapper/zkwrapper.cc

```diff
@@ -0,0 +1,534 @@
+	LOG(INFO) << CLASS_NAME <<  "creating seq ZNode at " << prepend_zk_root(path);
+
+	int len = prepend_zk_root(path).size();
+	new_path.resize(MAX_PATH_LEN);
+	int rc = zoo_create(zh,
+			prepend_zk_root(path).c_str(),
+			reinterpret_cast<const char *>(data.data()),
+			data.size(),
+			&ZOO_OPEN_ACL_UNSAFE,
+			flag,
+			reinterpret_cast<char *>(&new_path[0]),
+			MAX_PATH_LEN);
+	error_code = rc;
+	if (rc) { // Z_OK is 0, so if we receive anything else fail
+		LOG(ERROR) << CLASS_NAME <<  "Create for " << prepend_zk_root(path) << " failed " << rc;
+		print_error(error_code);
+		return false;
+	}
+	int i = 0;
+	LOG(INFO) << CLASS_NAME <<  "NEW path is " << new_path;
+	new_path.resize(len+10);
+	new_path = removeZKRoot(new_path);
+	LOG(INFO) << CLASS_NAME <<  "NEW path is now this" << new_path;
+	return true;
+}
+
+bool ZKWrapper::recursive_create(const std::string &path,
+		const std::vector <std::uint8_t> &data,
+		int &error_code) const {
+	for (int i=1; i<path.length(); ++i){
+		if (path[i] == '/'){
+			LOG(INFO) << CLASS_NAME <<  "Generating " << path.substr(0, i);
+			if (!create(path.substr(0, i), ZKWrapper::EMPTY_VECTOR, error_code)){
+				if (error_code != ZNODEEXISTS){
+					LOG(ERROR) << CLASS_NAME <<  "Failed to recursively create " << path;
+					print_error(error_code);
+					return false;
+				}
+			}
+			error_code = ZOK;
+		}
+	}
+	LOG(INFO) << CLASS_NAME <<  "Generating " << path;
+	return create(path, data, error_code);
+
+}
+
+bool ZKWrapper::wget(const std::string &path,
+		std::vector <std::uint8_t> &data,
+		watcher_fn watch,
+		void *watcherCtx,
+		int &error_code) const {
+	// TODO: Make this a constant value. Define a smarter retry policy for oversized data
+	int len = 0;
+	struct Stat stat;
+	error_code = zoo_wget(zh,
+			prepend_zk_root(path).c_str(),
+			watch,
+			watcherCtx,
+			reinterpret_cast<char *>(data.data()),
+			&len,
+			&stat);
+	if (error_code != ZOK) {
+		LOG(ERROR) << CLASS_NAME <<  "wget on " << path << " failed";
+		print_error(error_code);
+		return false;
+	}
+	data.resize(len);
+	return true;
+}
+
+bool ZKWrapper::get(const std::string &path,
+		std::vector <std::uint8_t> &data,
+		int &error_code) const {
+
+	// TODO: Make this a constant value. Define a smarter retry policy for oversized data
+	struct Stat stat;
+	int len = MAX_PAYLOAD;
+	// TODO: Perhaps we can be smarter about this
+	// LOG(INFO) << CLASS_NAME <<  "Data resizing to " << len;
+	data.resize(len);
+	// LOG(INFO) << CLASS_NAME <<  "Data resizing to 1;" << data.size();
+	error_code = zoo_get(zh,
+			prepend_zk_root(path).c_str(),
+			0,
+			reinterpret_cast<char *>(data.data()),
+			&len,
+			&stat);
+	if (error_code != ZOK) {
+		LOG(ERROR) << CLASS_NAME <<  "get on " << path << " failed";
+		print_error(error_code);
+		return false;
+	}
+	data.resize(len);
+	return true;
+}
+
+bool ZKWrapper::set(const std::string &path,
+		const std::vector <std::uint8_t> &data,
+		int &error_code,
+		int version) const {
```

Comment:
It seems that this is not possible. Version has a default value of -1, when I changed the order, I can't get my code compiled.
reference:
[http://stackoverflow.com/questions/5740296/missing-default-argument-compiler-error]